xdsl/dialects/arith.py

```diff
@@ -551,6 +455,9 @@ def get(operand1: Union[Operation, SSAValue],
             result_types=[IntegerType.from_width(1)],
             attributes={"predicate": IntegerAttr.from_int_and_width(arg, 64)})
 
+    def verify_(self) -> None:
```

Comment:
I would count comparisons as binary operations as well, considering they have two inputs. What exactly is a binary operation in this context? Aren't you describing a `BinarySameInputSameResultOperation`?

I think we might want to split this into several interfaces (or traits probably). The `Binary`, the `SameInputTypes`, and the `SameResultTypes` traits or something.

Reply:
I like your proposed approach. One of the ideas I have is the following.
We should have a different dialect that can be called relational.py.
Have a look here if you have time: https://github.com/devitocodes/devito/blob/master/devito/types/relational.py

Reply:
Yes I agree with this structure! Though notice that there are hard problems here due to our usage of decorators. AFAIK, the trait should not have a decorator atm.

Reply:
Maybe here you did not notice that` cmpi` is not inheriting from BinaryOperation!
It is as before, so we could skip for the moment.